docs/src/tutorials/entry-tutorial.md

```diff
@@ -0,0 +1,1085 @@
+a detailed explanation in the section on [Assignments in Apalache][].
+
+
+TODO: add the link to BinSearch4.tla and MC4_8.tla when we have them on github.
+
+## Step 5: Checking Postcondition for plenty of inputs
+
+Copy `BinSearch4.tla` to `BinSearch5.tla` and open `BinSearch5.tla`.
+
+In this step, we are going to check the invariant `Postcondition` for all
+possible input sequences and all input keys (for a fixed bit-width).
+
+Create the file `MC5_8.tla` with the following contents:
+
+```tla
+{{#include ../../../test/tla/bin-search/MC5_8.tla::34}}
+==================
+```
+
+Note that we introduced `INPUT_SEQ` and `INPUT_KEY` as parameters again. We
+cannot check `MC5_8.tla` just like that. If we try to check `MC5_8.tla`,
+Apalache would complain again about a value missing for `INPUT_SEQ`.
+
+To check the invariant for all sequences, we will use two advanced features
+of Apalache: [ConstInit predicate][] and [Value generators][].
+
+**ConstInit.** This idiom allows us to initialize `CONSTANTS` with a TLA+
+formula. Let us introduce the following operator definition in `MC5_8.tla`:
+
+```tla
+ConstInit ==
+    /\ INPUT_KEY \in Int
+    \* Seq(Int) is a set of all sequences that have integers as elements
+    /\ INPUT_SEQ \in Seq(Int) 
+```
+
+This is straightforward definition. However, it does not work in Apalache:
+
+```sh
+$ apalache-mc check --cinit=ConstInit --inv=Postcondition MC5_8.tla
+...
+MC5_8.tla:39:22-39:29: unsupported expression: Seq(_) produces an infinite set of unbounded sequences.
+Checker has found an error
+...
+```
+
+The issue with our definition of `ConstInit` is that it requires the model
+checker to reason about the infinite set of sequences, namely, `Seq(Int)`.
+Interestingly, the model checking does not complain about the expression
+`INPUT_KEY \in Int`. The reason is that this expression requires the model
+checker to reason about one integer, though it ranges over the infinite set of
+integers.
+
+**Value generators.** Fortunately, this problem can be easily circumvented by
+using Apalache [Value generators][], which are inspired by property-based testing,
+e.g., [QuickCheck][].
+
+Let us rewrite `ConstInit` in `MC5_8.tla` as follows:
+
+```tla
+ConstInit ==
+    /\ INPUT_KEY = Gen(1)
+    /\ INPUT_SEQ = Gen(MAX_INT)
+```
+
+In this new version, we use value generators to:
+
+  - produce an unrestricted integer to be used as a value of `INPUT_KEY` and
+  - produce a sequence of integers to be used as a value of `INPUT_SEQ`. This
+    sequence is unrestricted, except its length is bounded with `MAX_INT`,
+    which is exactly what we need in our case study.
+
+In contrast to property-based testing, an Apalache generator is not randomly
+producing values. Rather, it simply introduces a data structure whose size is
+bounded with the argument of `Gen`. This lets Apalache check all instances of
+the data structure, without enumerating the instances!
```

Comment:
Never having used value generators, I'm not clear on their syntax nor semantics from this explanation.

Reply:
Explained.